src/full/Agda/TypeChecking/Monad/Base.hs

```diff
@@ -2541,16 +2541,13 @@ shouldReduceDef f = asksTC envReduceDefs <&> \case
   OnlyReduceDefs defs -> f `Set.member` defs
   DontReduceDefs defs -> not $ f `Set.member` defs
 
-instance Semigroup ReduceDefs where
-  OnlyReduceDefs qs1 <> OnlyReduceDefs qs2 = OnlyReduceDefs $ Set.intersection qs1 qs2
-  OnlyReduceDefs qs1 <> DontReduceDefs qs2 = OnlyReduceDefs $ Set.difference   qs1 qs2
-  DontReduceDefs qs1 <> OnlyReduceDefs qs2 = OnlyReduceDefs $ Set.difference   qs2 qs1
-  DontReduceDefs qs1 <> DontReduceDefs qs2 = DontReduceDefs $ Set.union        qs1 qs2
```

Comment:
I see you opted to remove the old `{only/dont}ReduceDefs` API.  This warrants a migration guide.
An alternative would be to leave it but deprecate it.

Reply:
Yes, it's redundant with the new primitives (except that the set operations should be a bit cheaper on the Haskell side). I'm happy to either re-implement these as functions on the Agda side, or to write a migration guide. Also, the type of `withReconstructed` is changed by this PR, with `withReconstructed true` being the old behaviour, so it's a breaking change either way. I suppose adjusting for the new `withReconstructed` is very easy, while the other ones are a bit more work.

Reply:
Maybe some users of reflection could chime in on whether the old API to ReduceDefs should be available still.  @jespercockx @L-TChen

Reply:
I am not sure about what to do with `withReconstructed`. In general, it will be nice to re-implement old functions (using the same names) in `Agda.Builtin.Reflection` by the new primitives but deprecate them. So, the old ones are removed from the codebase, but they are still usable with a deprecation warning.

Reply:
One option is perhaps to put the reimplementations in the standard library.

Reply:
Many Agda users, including myself, do not base their development on the "standard" library, and there are other choices like Ulf's `agda-prelude`. Removing old primitives from `Agda.Builtin.Reflection` directly means that there is no transition period at all for other libraries. I think we have learned this lesson from https://github.com/agda/cubical/issues/743.

Reply:
I should have written that, if the reimplementation uses something like AVL trees (for performance reasons), then it might make sense to put it in the standard library.

Reply:
Oh, right. I agree.

Reply:
Ok, then I'll add functions that implement the two `reduceDefs` in `Agda.Builtin.Reflection` with a deprecation warning. If someone needs functions with better performance (which would surprise me) they'll have to do that on their own.
The problem with `withReconstructed` is of course that it has the same name as the old primitive. I'm not sure if it helps to add something like `withReconstructed' = withReconstructed true`.

Reply:
I've done this now. It's a bit annoying that I needed to define a bunch of functions to get the old behaviour, but I guess there's no way around that.